exploration/data-driven-tests.md

```diff
@@ -0,0 +1,243 @@
+# Data-driven tests
+
+Status: **Proposed**
+
+<details>
+	<summary>Metadata</summary>
+	<dl>
+		<dt>Contributors</dt>
+		<dd>@mradbourne</dd>
+		<dt>First proposed</dt>
+		<dd>2023-09-28</dd>
+		<dt>Pull Request</dt>
+		<dd>#495</dd>
+	</dl>
+</details>
+
+## Objective
+
+Establish a suitable approach for testing that different implmentations of MessageFormat 2 align with the specification.
+
+This approach includes _how_ tests are written: They should be captured in a single platform-agnostic format that can be utilized by all MF2 implementations. There should be no need to rewrite individual test cases for each platform.
+
+This approach also includes _what_ kind of tests are written. We need to identify which parts of MF2 should be covered by different types of test as a minimum.
+
+## Background
+
+Several pre-existing test files have been considered before forming this proposal:
+
+- [Unicode's Data Driven Test framework](https://github.com/unicode-org/conformance)
+- [message-format-wg XML test format](https://github.com/unicode-org/message-format-wg/tree/514758923abac13a2c5eb71b6b6cdef4a181280e/test)
+- [Intl.MessageFormat polyfill tests](https://github.com/messageformat/messageformat/tree/main/packages/mf2-messageformat/src)
+- [ICU4J test examples](https://github.com/unicode-org/icu/tree/main/icu4j/main/core/src/test/java/com/ibm/icu/dev/test/message2)
+- [Tim Chevalier's draft ICU4C tests](https://github.com/catamorphism/icu/blob/parser-plus-data-model-plus-full-api/icu4c/source/test/intltest/messageformat2test.cpp)
+
+## Use-Cases
+
+The main platforms for which the tests should initially run are:
+
+- Node.js
+- ICU4J (Java)
+- ICU4C (C++)
+
+Other platforms, such as ICU4X (Rust) may be added later.
+
+## Requirements
+
+- Test framework
+
+  - The test cases and assertions must be easy to read.
+  - The test cases and assertions must be completely platform-agnostic.
+  - The framework must include the platform-specific test executors as part of the solution.
+  - The framework must be extendable with new executors (e.g. ICU4X) and it should be clear how to do this.
+
+- Test content
+  - Syntax tests: Testing which standard registry functions are invoked when different arguments are passed to a given pattern.
```

Comment:
I'm not sure this is the right description of "syntax tests"? Patterns might not invoke any functions, but we would still want to test that correct patterns don't generate syntax errors and that incorrect patterns generate the right syntax errors.

Reply:
Good point. I've updated accordingly.